pkg/cloud/services/compute/instance.go

```diff
@@ -482,47 +381,83 @@ func getServerNetworks(networkClient *gophercloud.ServiceClient, networkParams [
+	allPages, err := ports.List(is.networkClient, ports.ListOpts{
+		Name:      portName,
+		NetworkID: net.ID,
+	}).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("searching for existing port for server: %v", err)
 	}
-	for _, element := range list {
-		if element == name {
-			return true
-		}
+	portList, err := ports.ExtractPorts(allPages)
+	if err != nil {
+		return nil, fmt.Errorf("searching for existing port for server err: %v", err)
+	}
+
+	if len(portList) != 0 {
+		return &portList[0], nil
 	}
-	return false
-}
 
-func createPort(is *Service, clusterName string, name string, net *infrav1.Network, securityGroups *[]string) (ports.Port, error) {
 	portCreateOpts := ports.CreateOpts{
-		Name:           name,
+		Name:           portName,
 		NetworkID:      net.ID,
 		SecurityGroups: securityGroups,
 		Description:    fmt.Sprintf("Created by cluster-api-provider-openstack cluster %s", clusterName),
 	}
 	if net.Subnet.ID != "" {
 		portCreateOpts.FixedIPs = []ports.IP{{SubnetID: net.Subnet.ID}}
 	}
-	newPort, err := ports.Create(is.networkClient, portCreateOpts).Extract()
+	port, err := ports.Create(is.networkClient, portCreateOpts).Extract()
 	if err != nil {
-		return ports.Port{}, fmt.Errorf("create port for server: %v", err)
+		record.Warnf(object, "FailedCreatePort", "Failed to create port %s: %v", portName, err)
+		return nil, err
 	}
-	return *newPort, nil
+
+	record.Eventf(object, "SuccessfulCreatePort", "Created port %s with id %s", port.Name, port.ID)
+	return port, nil
 }
 
-func deletePorts(s *Service, nets []servers.Network) error {
-	for _, n := range nets {
-		_, err := ports.Get(s.networkClient, n.Port).Extract()
-		if err != nil {
-			if capoerrors.IsNotFound(err) {
-				return nil
-			}
-		}
-		if err := ports.Delete(s.networkClient, n.Port).ExtractErr(); err != nil {
-			return err
-		}
+func getOrCreateTrunk(is *Service, object runtime.Object, trunkName, portID string) (*trunks.Trunk, error) {
+	allPages, err := trunks.List(is.networkClient, trunks.ListOpts{
+		Name:   trunkName,
+		PortID: portID,
+	}).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("searching for existing trunk for server err: %v", err)
+	}
+	trunkList, err := trunks.ExtractTrunks(allPages)
+	if err != nil {
+		return nil, fmt.Errorf("searching for existing trunk for server err: %v", err)
+	}
+
+	if len(trunkList) != 0 {
+		return &trunkList[0], nil
+	}
+
+	trunkCreateOpts := trunks.CreateOpts{
+		Name:   trunkName,
+		PortID: portID,
+	}
+	trunk, err := trunks.Create(is.networkClient, trunkCreateOpts).Extract()
+	if err != nil {
+		record.Warnf(object, "FailedCreateTrunk", "Failed to create trunk %s: %v", trunkName, err)
+		return nil, err
+	}
+
+	record.Eventf(object, "SuccessfulCreateTrunk", "Created trunk %s with id %s", trunk.Name, trunk.ID)
+	return trunk, nil
+}
+
+func replaceAllAttributestag(is *Service, object runtime.Object, trunkID string, tags []string) error {
```

Comment:
Is this func is specific to trunks? If so I think we should add some context to the function name, otherwise we could rename `trunkID` to `resourceID` like it is named in the `ReplaceAll` func

Reply:
I want to express the context by function naming `replaceAllAttributestag` and argument name `trunkID`.

Reply:
I don't know if this would be a good pattern. If there later is a code review and someone does a `replaceAllAttributestag(is, object, port.ID, i.Tags)` the context would be gone.

Also:
If we keep the name, it should be called `AttributesTags` (upper-case T, and s), just the package name is all lowercase and inside the gophercloud package it is also called `AttributesTags`

Reply:
@chrischdi So you want to rename to `replaceAllAttributesTagOfTrunk` or something?
I agree with upper-case T. This function name should start with small letter, since this function is not public.

Reply:
I'd prefer something like that, however I would be also ok with `replaceAllAttributesTag` if this is still your favourite :relieved:

Reply:
Renamed `replaceAllAttributestag` to `replaceAllAttributesTags`. Sorry forgot to say ending with small s.